torch/distributed/algorithms/_comm_hooks/allreduce_hook.py

```diff
@@ -0,0 +1,59 @@
+import torch
+import torch.distributed as dist
+from torch.distributed import distributed_c10d
+
+
+class AllReduceState(object):
+    r"""
+    Stores parameters, needed to perform ``all_reduce`` algorithm
+    within a communication hook.
+
+    Args:
+        process_group (ProcessGroup): The process group to be used for all-reduce.
+        world_size (int): The number of nodes in a process group.
+            Determined based on a ``process_group``.
+        gradient_predivide_factor (float): A factor for gradients' pre-division.
+        gradient_postdivide_factor (float): A factor for gradients' post-division.
+    """
+
+    __slots__ = [
+        "process_group",
+        "world_size",
+        "gradient_predivide_factor",
+        "gradient_postdivide_factor"
+    ]
+
+    def __init__(
+        self,
+        process_group
+    ):
+        self.process_group = process_group if process_group is not None else distributed_c10d._get_default_group()
+        self.world_size = dist.get_world_size(process_group)
+        self.gradient_predivide_factor = self._get_gradient_predivide_factor(
+            self.world_size
+        )
+        self.gradient_postdivide_factor = self.world_size / self.gradient_predivide_factor
+
+    # setting two factors 'self.gradient_predivide_factor'
+    # and 'self.gradient_postdivide_factor' to avoid underflow and overflow
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
+        return float(factor)
+
+
+def allreduce_hook(state: AllReduceState, grad: torch.Tensor):
+    r"""
+    This FSDP communication hook implements ``all_reduce`` algorithm
+    and a neccessary post-division of gradients.
+
+    Args:
+        state (AllReduceState): State information, configures pre- and post-division factors
+        grad (torch.Tensor): A gradient for the local batch that needs to be communicated across ranks.
+    """
+    # `all_reduce` is a default hook and grad is already pre-divided
+    dist.all_reduce(grad, group=state.process_group)
```

Comment:
looks like this hook expects a pre-divided gradient - would be good to document this clearly, that we expect predivision, but perform postdivision (also tbh that might be a little confusing to users but it's fine as this hook simulates existing FSDP behavior, we can add an allreduce hook that does not postdivide later).

Consider the above, might also be good to rename this hook to `allreduce_with_postdivide_hook`?

Reply:
I've reconsidered the current behavior of a default `all_reduce` hook and in the updated version it takes care of pre- and post-division. This way there is no necessity in different `all_reduce` hooks and confusion is eliminated. Note: as was discussed earlier, `FSDP. _post_backward_hook` does not use sharding strategy explicitly for pre-division of gradients.